Source/Core/VideoBackends/Vulkan/Renderer.cpp

```diff
@@ -737,22 +737,133 @@ bool Renderer::DrawFrameDump(const EFBRectangle& rc, u32 xfb_addr,
+  FlushFrameDump();
+
+  // Remove the fence callback that we registered earlier, one less function that needs to be
+  // called when preparing a command buffer.
+  g_command_buffer_mgr->RemoveFencePointCallback(m_frame_dump_images.data());
+  m_frame_dumping_active = false;
+}
+
+void Renderer::OnFrameDumpImageReady(VkFence fence)
+{
+  for (FrameDumpImage& frame : m_frame_dump_images)
+  {
+    // fence being a null handle means that we don't have to wait to re-use this image.
+    if (frame.fence == fence)
+      frame.fence = VK_NULL_HANDLE;
+  }
+}
+
+void Renderer::WriteFrameDumpImage(size_t index)
+{
+  FrameDumpImage& frame = m_frame_dump_images[index];
+  _assert_(frame.pending);
+
+  // Check fence has been signaled.
+  // The callback here should set fence to null.
+  if (frame.fence != VK_NULL_HANDLE)
+  {
+    g_command_buffer_mgr->WaitForFence(frame.fence);
+    _assert_(frame.fence == VK_NULL_HANDLE);
+  }
+
+  // Copy the now-populated image data to the output file.
+  AVIDump::Frame state = AVIDump::FetchState(frame.ticks);
   FinishFrameData();
+  DumpFrameData(reinterpret_cast<const u8*>(frame.readback_texture->GetMapPointer()),
+                static_cast<int>(frame.readback_texture->GetWidth()),
+                static_cast<int>(frame.readback_texture->GetHeight()),
+                static_cast<int>(frame.readback_texture->GetRowStride()), state);
+
+  frame.pending = false;
+}
+
+StagingTexture2D* Renderer::PrepareFrameDumpImage(size_t index, u32 width, u32 height, u64 ticks)
+{
+  FrameDumpImage& image = m_frame_dump_images[index];
+
+  // If there is an image that hasn't been written to the frame dump yet, write it.
+  if (image.pending)
+    WriteFrameDumpImage(index);
+
+  // Ensure the dimensions of the readback texture are sufficient.
+  if (!image.readback_texture || width > image.readback_texture->GetWidth() ||
+      height > image.readback_texture->GetHeight())
+  {
+    // Allocate a new readback texture.
+    // The reset() call is here so that the memory is released before allocating the new texture.
+    image.readback_texture.reset();
+    image.readback_texture = StagingTexture2D::Create(STAGING_BUFFER_TYPE_READBACK, width, height,
+                                                      EFB_COLOR_TEXTURE_FORMAT);
+
+    if (!image.readback_texture || !image.readback_texture->Map())
+    {
+      // Not actually fatal, just means we can't dump this frame.
+      PanicAlert("Failed to allocate frame dump readback texture.");
+      image.readback_texture.reset();
+      return nullptr;
+    }
+  }
+
+  // The copy happens immediately after this function returns, so flag this frame as pending.
+  image.fence = g_command_buffer_mgr->GetCurrentCommandBufferFence();
+  image.ticks = ticks;
+  image.pending = true;
+
+  // This frame is queued, so use the next buffer for the next frame.
+  m_current_frame_dump_image = (m_current_frame_dump_image + 1) % FRAME_DUMP_BUFFERED_FRAMES;
```

Comment:
Am I correct here?
PrepareFrameDumpImage(m_current_frame_dump_image) may start to encode the image in the buffer m_current_frame_dump_image. Afterwards it returns this buffer and you'll copy by CopyFromImage into this buffer.

I think this line should be moved after WriteFrameDumpImage.